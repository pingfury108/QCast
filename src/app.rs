use async_trait::async_trait;
use loco_rs::{
    app::{AppContext, Hooks, Initializer},
    bgworker::Queue,
    boot::{create_app, BootResult, StartMode},
    config::Config,
    controller::AppRoutes,
    db::{self, truncate_table},
    environment::Environment,
    task::Tasks,
    Result,
};
use migration::Migrator;
use std::path::Path;

#[allow(unused_imports)]
use crate::{controllers, initializers, models::_entities::users, tasks, workers};

pub struct App;
#[async_trait]
impl Hooks for App {
    fn app_name() -> &'static str {
        env!("CARGO_CRATE_NAME")
    }

    fn app_version() -> String {
        format!(
            "{} ({})",
            env!("CARGO_PKG_VERSION"),
            option_env!("BUILD_SHA")
                .or(option_env!("GITHUB_SHA"))
                .unwrap_or("dev")
        )
    }

    async fn boot(
        mode: StartMode,
        environment: &Environment,
        config: Config,
    ) -> Result<BootResult> {
        create_app::<Self, Migrator>(mode, environment, config).await
    }

    async fn initializers(_ctx: &AppContext) -> Result<Vec<Box<dyn Initializer>>> {
        Ok(vec![])
    }

    fn routes(_ctx: &AppContext) -> AppRoutes {
        AppRoutes::with_default_routes() // controller routes below
            .add_route(controllers::books::routes())
            .add_route(controllers::chapters::routes())
            .add_route(controllers::medias::routes())
            .add_route(controllers::public::routes())
            .add_route(controllers::auth::routes())
            .add_route(controllers::dashboard::routes())
            // 后台管理路由
            .add_route(controllers::admin::users::routes())
            .add_route(controllers::admin::groups::routes())
    }
    async fn connect_workers(_ctx: &AppContext, _queue: &Queue) -> Result<()> {
        // 暂无 worker
        Ok(())
    }

    #[allow(unused_variables)]
    fn register_tasks(tasks: &mut Tasks) {
        tasks.register(tasks::create_superadmin::CreateSuperadmin);
        // tasks-inject (do not remove)
    }
    async fn truncate(ctx: &AppContext) -> Result<()> {
        truncate_table(&ctx.db, users::Entity).await?;
        Ok(())
    }
    async fn seed(ctx: &AppContext, base: &Path) -> Result<()> {
        db::seed::<users::ActiveModel>(&ctx.db, &base.join("users.yaml").display().to_string())
            .await?;
        Ok(())
    }
}
