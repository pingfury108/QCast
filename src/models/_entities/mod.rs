//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub mod prelude;

pub mod books;
pub mod chapters;
pub mod medias;
pub mod site_settings;
pub mod user_group_members;
pub mod user_groups;
pub mod users;
