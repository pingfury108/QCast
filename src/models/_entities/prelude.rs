//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

pub use super::books::Entity as Books;
pub use super::chapters::Entity as Chapters;
pub use super::medias::Entity as Medias;
pub use super::site_settings::Entity as SiteSettings;
pub use super::user_group_members::Entity as UserGroupMembers;
pub use super::user_groups::Entity as UserGroups;
pub use super::users::Entity as Users;
