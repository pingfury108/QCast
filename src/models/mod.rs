pub mod _entities;
pub mod books;
pub mod chapters;
pub mod medias;
pub mod user_group_members;
pub mod user_groups;
pub mod users;
