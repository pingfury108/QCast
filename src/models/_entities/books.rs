//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.16

use sea_orm::entity::prelude::*;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "books")]
pub struct Model {
    pub created_at: DateTimeWithTimeZone,
    pub updated_at: DateTimeWithTimeZone,
    #[sea_orm(primary_key)]
    pub id: i32,
    pub title: String,
    #[sea_orm(column_type = "Text", nullable)]
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
    pub user_id: i32,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}
