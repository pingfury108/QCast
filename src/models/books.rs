pub use super::_entities::books::{ActiveModel, Column, Entity, Model};
pub use super::_entities::users;
use sea_orm::entity::prelude::*;
use sea_orm::sea_query::{Condition, Expr, Func};
use sea_orm::QueryOrder;
use serde::{Deserialize, Serialize};
pub type Books = Entity;

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    async fn before_save<C>(self, _db: &C, insert: bool) -> std::result::Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        let mut this = self;

        if insert {
            // 设置默认值
            if this.sort_order.is_unchanged() {
                this.sort_order = sea_orm::ActiveValue::Set(Some(0));
            }
            if this.is_public.is_unchanged() {
                this.is_public = sea_orm::ActiveValue::Set(Some(false));
            }
        }

        if !insert && this.updated_at.is_unchanged() {
            this.updated_at = sea_orm::ActiveValue::Set(chrono::Utc::now().into());
        }

        Ok(this)
    }
}

// implement your read-oriented logic here
impl Model {
    /// 获取用户的所有书籍
    pub async fn find_by_user(db: &DatabaseConnection, user_id: i32) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::UserId.eq(user_id))
            .all(db)
            .await
    }

    /// 获取顶级书籍（parent_id 为 null）
    pub async fn find_root_books(
        db: &DatabaseConnection,
        user_id: i32,
    ) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::UserId.eq(user_id))
            .filter(Column::ParentId.is_null())
            .order_by_asc(Column::SortOrder)
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 获取子书籍
    pub async fn find_children(
        db: &DatabaseConnection,
        parent_id: i32,
    ) -> Result<Vec<Model>, DbErr> {
        Entity::find()
            .filter(Column::ParentId.eq(parent_id))
            .order_by_asc(Column::SortOrder)
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 搜索书籍
    pub async fn search(
        db: &DatabaseConnection,
        user_id: i32,
        query: &str,
    ) -> Result<Vec<Model>, DbErr> {
        // 转换为小写进行大小写不敏感搜索（跨数据库兼容）
        let query_lower = query.to_lowercase();
        let search_pattern = format!("%{query_lower}%");

        Entity::find()
            .filter(Column::UserId.eq(user_id))
            .filter(
                Condition::any()
                    .add(Expr::expr(Func::lower(Expr::col(Column::Title))).like(&search_pattern))
                    .add(
                        Expr::expr(Func::lower(Expr::col(Column::Description)))
                            .like(&search_pattern),
                    ),
            )
            .order_by_asc(Column::SortOrder)
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await
    }

    /// 搜索书籍（包含所有父书籍）
    pub async fn search_with_parents(
        db: &DatabaseConnection,
        user_id: i32,
        query: &str,
    ) -> Result<Vec<Model>, DbErr> {
        // 先获取匹配的书籍
        let matched_books = Self::search(db, user_id, query).await?;

        if matched_books.is_empty() {
            return Ok(Vec::new());
        }

        // 收集所有需要包含的书籍ID
        let mut book_ids = std::collections::HashSet::new();
        for book in &matched_books {
            book_ids.insert(book.id);

            // 递归查找所有父书籍
            let mut current_parent_id = book.parent_id;
            while let Some(parent_id) = current_parent_id {
                if book_ids.contains(&parent_id) {
                    break; // 避免循环
                }
                book_ids.insert(parent_id);

                // 查找父书籍
                if let Some(parent) = Entity::find_by_id(parent_id)
                    .filter(Column::UserId.eq(user_id))
                    .one(db)
                    .await?
                {
                    current_parent_id = parent.parent_id;
                } else {
                    break;
                }
            }
        }

        // 获取所有书籍（包括匹配的和父书籍）
        let all_books = Entity::find()
            .filter(Column::UserId.eq(user_id))
            .filter(Column::Id.is_in(book_ids.into_iter().collect::<Vec<_>>()))
            .order_by_asc(Column::SortOrder)
            .order_by_asc(Column::CreatedAt)
            .all(db)
            .await?;

        Ok(all_books)
    }

    /// 获取书籍的完整树形结构（递归获取所有子书籍）
    pub async fn get_tree(db: &DatabaseConnection, book_id: i32) -> Result<BookTree, DbErr> {
        let book = Entity::find_by_id(book_id)
            .one(db)
            .await?
            .ok_or(DbErr::RecordNotFound(format!("Book {} not found", book_id)))?;

        let children = Self::find_children(db, book_id).await?;
        let mut tree_children = Vec::new();

        // 递归获取子书籍的树形结构
        for child in children {
            // 使用 Box::pin 来避免递归 async fn 问题
            let child_tree_future = Box::pin(Self::get_tree(db, child.id));
            let child_tree = child_tree_future.await?;
            tree_children.push(child_tree);
        }

        Ok(BookTree {
            book,
            children: tree_children,
        })
    }

    /// 获取用户的所有书籍的树形结构
    pub async fn get_user_tree(
        db: &DatabaseConnection,
        user_id: i32,
    ) -> Result<Vec<BookTree>, DbErr> {
        let root_books = Self::find_root_books(db, user_id).await?;
        let mut tree = Vec::new();

        for book in root_books {
            let book_tree = Self::get_tree(db, book.id).await?;
            tree.push(book_tree);
        }

        Ok(tree)
    }
}

/// 书籍树形结构
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BookTree {
    pub book: Model,
    pub children: Vec<BookTree>,
}

// implement your write-oriented logic here
impl ActiveModel {}

// implement your custom finders, selectors oriented logic here
impl Entity {}
