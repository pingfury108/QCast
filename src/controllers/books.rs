#![allow(clippy::missing_errors_doc)]
#![allow(clippy::unnecessary_struct_initialization)]
#![allow(clippy::unused_async)]
use axum::debug_handler;
use axum::extract::Query;
use axum::routing::method_routing::delete as axum_delete;
use loco_rs::prelude::*;
use serde::{Deserialize, Serialize};

use crate::models::_entities::books::{ActiveModel, Column, Entity, Model};
use crate::models::users;
use crate::views::books::{BookResponse, BookTreeResponse};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateParams {
    pub title: String,
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateParams {
    pub title: Option<String>,
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ReorderParams {
    pub sort_order: i32,
}

async fn load_item(ctx: &AppContext, id: i32, user_id: i32) -> Result<Model> {
    let item = Entity::find_by_id(id)
        .filter(Column::UserId.eq(user_id))
        .one(&ctx.db)
        .await?;
    item.ok_or_else(|| Error::NotFound)
}

/// 获取当前用户的所有书籍
#[debug_handler]
pub async fn list(auth: auth::JWT, State(ctx): State<AppContext>) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    let books = Entity::find()
        .filter(Column::UserId.eq(user.id))
        .all(&ctx.db)
        .await?;

    let responses: Vec<BookResponse> = books.into_iter().map(BookResponse::from).collect();

    format::json(responses)
}

/// 创建书籍
#[debug_handler]
pub async fn create(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Json(params): Json<CreateParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    let item = ActiveModel {
        user_id: Set(user.id),
        title: Set(params.title),
        description: Set(params.description),
        cover_image: Set(params.cover_image),
        parent_id: Set(params.parent_id),
        sort_order: Set(params.sort_order),
        is_public: Set(params.is_public),
        ..Default::default()
    };

    let item = item.insert(&ctx.db).await?;

    format::json(BookResponse::from(item))
}

/// 获取书籍详情
#[debug_handler]
pub async fn show(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    format::json(BookResponse::from(item))
}

/// 更新书籍
#[debug_handler]
pub async fn update(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<UpdateParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    let mut item = item.into_active_model();

    if let Some(title) = params.title {
        item.title = Set(title);
    }
    if let Some(description) = params.description {
        item.description = Set(Some(description));
    }
    if let Some(cover_image) = params.cover_image {
        item.cover_image = Set(Some(cover_image));
    }
    if let Some(parent_id) = params.parent_id {
        item.parent_id = Set(Some(parent_id));
    }
    if let Some(sort_order) = params.sort_order {
        item.sort_order = Set(Some(sort_order));
    }
    if let Some(is_public) = params.is_public {
        item.is_public = Set(Some(is_public));
    }

    let item = item.update(&ctx.db).await?;

    format::json(BookResponse::from(item))
}

/// 删除书籍
#[debug_handler]
pub async fn delete(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    item.delete(&ctx.db).await?;
    format::empty()
}

/// 搜索书籍
#[debug_handler]
pub async fn search(
    auth: auth::JWT,
    State(ctx): State<AppContext>,
    Query(params): Query<std::collections::HashMap<String, String>>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    let query = params.get("q").map(|q| q.as_str()).unwrap_or("");

    if query.is_empty() {
        return format::json(Vec::<BookResponse>::new());
    }

    let books = Model::search(&ctx.db, user.id, query).await?;
    let responses: Vec<BookResponse> = books.into_iter().map(BookResponse::from).collect();

    format::json(responses)
}

/// 获取书籍的完整树形结构
#[debug_handler]
pub async fn tree(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;

    // 验证用户是否有权限访问该书籍
    let _ = load_item(&ctx, id, user.id).await?;

    let book_tree = Model::get_tree(&ctx.db, id).await?;
    let tree_response = BookTreeResponse::from(book_tree);
    format::json(tree_response)
}

/// 调整书籍顺序
#[debug_handler]
pub async fn reorder(
    auth: auth::JWT,
    Path(id): Path<i32>,
    State(ctx): State<AppContext>,
    Json(params): Json<ReorderParams>,
) -> Result<Response> {
    let user = users::Model::find_by_pid(&ctx.db, &auth.claims.pid).await?;
    let item = load_item(&ctx, id, user.id).await?;

    let mut item = item.into_active_model();
    item.sort_order = Set(Some(params.sort_order));
    let item = item.update(&ctx.db).await?;

    format::json(BookResponse::from(item))
}

pub fn routes() -> Routes {
    Routes::new()
        .prefix("api/books/")
        .add("/", get(list))
        .add("/", post(create))
        .add("/search", get(search))
        .add("/{id}", get(show))
        .add("/{id}", put(update))
        .add("/{id}", patch(update))
        .add("/{id}", axum_delete(delete))
        .add("/{id}/tree", get(tree))
        .add("/{id}/reorder", post(reorder))
}
