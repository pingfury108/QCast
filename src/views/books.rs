use crate::models::_entities::books::Model;
use crate::models::books::BookTree;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct BookResponse {
    pub id: i32,
    pub title: String,
    pub description: Option<String>,
    pub cover_image: Option<String>,
    pub parent_id: Option<i32>,
    pub sort_order: Option<i32>,
    pub is_public: Option<bool>,
    pub user_id: i32,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

impl From<Model> for BookResponse {
    fn from(book: Model) -> Self {
        Self {
            id: book.id,
            title: book.title,
            description: book.description,
            cover_image: book.cover_image,
            parent_id: book.parent_id,
            sort_order: book.sort_order,
            is_public: book.is_public,
            user_id: book.user_id_id,
            created_at: book.created_at.into(),
            updated_at: book.updated_at.into(),
        }
    }
}

/// 书籍树形结构响应
#[derive(Debug, Serialize, Deserialize)]
pub struct BookTreeResponse {
    pub book: BookResponse,
    pub children: Vec<BookTreeResponse>,
}

impl From<BookTree> for BookTreeResponse {
    fn from(tree: BookTree) -> Self {
        Self {
            book: BookResponse::from(tree.book),
            children: tree
                .children
                .into_iter()
                .map(BookTreeResponse::from)
                .collect(),
        }
    }
}
