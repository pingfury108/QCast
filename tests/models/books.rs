use insta::{assert_debug_snapshot, with_settings};
use loco_rs::testing::prelude::*;
use qcast::app::App;
use qcast::models::_entities::books::{ActiveModel, Entity};
use qcast::models::books;
use sea_orm::{
    ActiveModelTrait, ColumnTrait, EntityTrait, PaginatorTrait, QueryFilter, QueryOrder, Set,
};
use serial_test::serial;

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_create_book() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    let book_data = ActiveModel {
        title: Set("测试书籍".to_string()),
        description: Set(Some("这是一本测试书籍".to_string())),
        user_id: Set(1), // 假设用户ID为1
        ..Default::default()
    };

    let created_book = book_data.insert(&boot.app_context.db).await.unwrap();

    assert!(created_book.id > 0);
    assert_eq!(created_book.title, "测试书籍");
    assert_eq!(
        created_book.description,
        Some("这是一本测试书籍".to_string())
    );
    assert_eq!(created_book.user_id, 1);

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(created_book);
    });
}

#[tokio::test]
#[serial]
async fn can_find_books_by_user() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建两本书，一本属于用户1，一本属于用户2
    let _book1 = ActiveModel {
        title: Set("用户1的书".to_string()),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let _book2 = ActiveModel {
        title: Set("用户2的书".to_string()),
        user_id: Set(2),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 查询用户1的书籍
    let user1_books = Entity::find()
        .filter(books::Column::UserId.eq(1))
        .all(&boot.app_context.db)
        .await
        .unwrap();

    assert_eq!(user1_books.len(), 1);
    assert_eq!(user1_books[0].title, "用户1的书");

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(user1_books);
    });
}

#[tokio::test]
#[serial]
async fn can_get_book_tree() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建根书籍
    let root_book = ActiveModel {
        title: Set("根书籍".to_string()),
        user_id: Set(1),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 创建子书籍
    let child_book = ActiveModel {
        title: Set("子书籍".to_string()),
        user_id: Set(1),
        parent_id: Set(Some(root_book.id)),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 创建孙子书籍
    let grandchild_book = ActiveModel {
        title: Set("孙子书籍".to_string()),
        user_id: Set(1),
        parent_id: Set(Some(child_book.id)),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 查询所有书籍并验证树结构
    let all_books = Entity::find()
        .filter(books::Column::UserId.eq(1))
        .order_by_asc(books::Column::SortOrder)
        .all(&boot.app_context.db)
        .await
        .unwrap();

    assert_eq!(all_books.len(), 3);

    // 验证父子关系
    let child = all_books.iter().find(|b| b.id == child_book.id).unwrap();
    assert_eq!(child.parent_id, Some(root_book.id));

    let grandchild = all_books
        .iter()
        .find(|b| b.id == grandchild_book.id)
        .unwrap();
    assert_eq!(grandchild.parent_id, Some(child_book.id));

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(all_books);
    });
}

#[tokio::test]
#[serial]
async fn can_search_books() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建多本书籍
    ActiveModel {
        title: Set("编程指南".to_string()),
        description: Set(Some("学习编程的完整指南".to_string())),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    ActiveModel {
        title: Set("设计模式".to_string()),
        description: Set(Some("软件设计模式详解".to_string())),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    ActiveModel {
        title: Set("算法导论".to_string()),
        description: Set(Some("算法与数据结构".to_string())),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 搜索包含"设计"的书籍
    let search_results = Entity::find()
        .filter(
            books::Column::Title
                .contains("设计")
                .or(books::Column::Description.contains("设计")),
        )
        .all(&boot.app_context.db)
        .await
        .unwrap();

    assert_eq!(search_results.len(), 1);
    assert_eq!(search_results[0].title, "设计模式");

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(search_results);
    });
}

#[tokio::test]
#[serial]
async fn can_search_books_with_parents() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建书籍层级结构
    let root_book = ActiveModel {
        title: Set("根书籍".to_string()),
        user_id: Set(1),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let child_book = ActiveModel {
        title: Set("子书籍".to_string()),
        user_id: Set(1),
        parent_id: Set(Some(root_book.id)),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let grandchild_book = ActiveModel {
        title: Set("孙子书籍目标".to_string()),
        description: Set(Some("这是目标书籍".to_string())),
        user_id: Set(1),
        parent_id: Set(Some(child_book.id)),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 创建一个无关的书籍
    let _other_book = ActiveModel {
        title: Set("无关书籍".to_string()),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 使用 search_with_parents 搜索包含"目标"的书籍
    let search_results = books::Model::search_with_parents(&boot.app_context.db, 1, "目标")
        .await
        .unwrap();

    // 应该返回孙子书籍、子书籍和根书籍（3个）
    assert_eq!(search_results.len(), 3);

    // 验证包含所有祖先
    let ids: Vec<i32> = search_results.iter().map(|b| b.id).collect();
    assert!(ids.contains(&root_book.id));
    assert!(ids.contains(&child_book.id));
    assert!(ids.contains(&grandchild_book.id));

    // 验证不包含无关书籍
    assert!(!ids.contains(&_other_book.id));

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(search_results);
    });
}

#[tokio::test]
#[serial]
async fn can_update_book() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建书籍
    let book = ActiveModel {
        title: Set("原始标题".to_string()),
        description: Set(Some("原始描述".to_string())),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 更新书籍
    let mut book_to_update: ActiveModel = book.into();
    book_to_update.title = Set("更新后的标题".to_string());
    book_to_update.description = Set(Some("更新后的描述".to_string()));

    let updated_book = book_to_update.update(&boot.app_context.db).await.unwrap();

    assert_eq!(updated_book.title, "更新后的标题");
    assert_eq!(updated_book.description, Some("更新后的描述".to_string()));
    assert!(updated_book.updated_at > updated_book.created_at);

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(updated_book);
    });
}

#[tokio::test]
#[serial]
async fn can_delete_book() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建书籍
    let book = ActiveModel {
        title: Set("待删除的书籍".to_string()),
        user_id: Set(1),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let book_id = book.id;

    // 删除书籍
    Entity::delete_by_id(book_id)
        .exec(&boot.app_context.db)
        .await
        .unwrap();

    // 验证书籍已被删除
    let deleted_book = Entity::find_by_id(book_id)
        .one(&boot.app_context.db)
        .await
        .unwrap();
    assert!(deleted_book.is_none());
}

#[tokio::test]
#[serial]
async fn can_handle_book_hierarchy() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建多层级的书籍结构
    let level1 = ActiveModel {
        title: Set("第一级".to_string()),
        user_id: Set(1),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let _level2_1 = ActiveModel {
        title: Set("第二级-第一章".to_string()),
        user_id: Set(1),
        parent_id: Set(Some(level1.id)),
        sort_order: Set(Some(1)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    let _level2_2 = ActiveModel {
        title: Set("第二级-第二章".to_string()),
        user_id: Set(1),
        parent_id: Set(Some(level1.id)),
        sort_order: Set(Some(2)),
        ..Default::default()
    }
    .insert(&boot.app_context.db)
    .await
    .unwrap();

    // 查询第一级书籍的所有子书籍
    let children = Entity::find()
        .filter(books::Column::ParentId.eq(Some(level1.id)))
        .order_by_asc(books::Column::SortOrder)
        .all(&boot.app_context.db)
        .await
        .unwrap();

    assert_eq!(children.len(), 2);
    assert_eq!(children[0].title, "第二级-第一章");
    assert_eq!(children[1].title, "第二级-第二章");

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(children);
    });
}

#[tokio::test]
#[serial]
async fn can_paginate_books() {
    configure_insta!();

    let boot = boot_test::<App>().await.unwrap();
    seed::<App>(&boot.app_context).await.unwrap();

    // 创建多本书籍用于分页测试
    for i in 1..=10 {
        ActiveModel {
            title: Set(format!("书籍 {:02}", i)),
            user_id: Set(1),
            ..Default::default()
        }
        .insert(&boot.app_context.db)
        .await
        .unwrap();
    }

    // 分页查询：第1页，每页3条
    let page1 = Entity::find()
        .filter(books::Column::UserId.eq(1))
        .order_by_asc(books::Column::Title)
        .paginate(&boot.app_context.db, 3);

    let page1_results = page1.fetch_page(0).await.unwrap();
    assert_eq!(page1_results.len(), 3);

    // 分页查询：第2页，每页3条
    let page2_results = page1.fetch_page(1).await.unwrap();
    assert_eq!(page2_results.len(), 3);

    // 验证排序正确
    assert_eq!(page1_results[0].title, "书籍 01");
    assert_eq!(page1_results[1].title, "书籍 02");
    assert_eq!(page1_results[2].title, "书籍 03");
    assert_eq!(page2_results[0].title, "书籍 04");

    with_settings!({
        filters => cleanup_book_model()
    }, {
        assert_debug_snapshot!(page1_results);
    });
}

fn cleanup_book_model() -> Vec<(&'static str, &'static str)> {
    vec![
        (r"id: \d+,", "id: ID"),
        (r"user_id: \d+,", "user_id: USER_ID"),
        (r"parent_id: Some\(\d+\)", "parent_id: Some(PARENT_ID)"),
        (
            r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?\+\d{2}:\d{2}",
            "DATE",
        ),
    ]
}
