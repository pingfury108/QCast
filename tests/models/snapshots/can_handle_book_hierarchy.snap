---
source: tests/models/books.rs
expression: children
---
[
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        title: "第二级-第一章",
        description: None,
        cover_image: None,
        parent_id: Some(
            1,
        ),
        sort_order: Some(
            1,
        ),
        is_public: None,
        user_id: ID
    },
    Model {
        created_at: DATE,
        updated_at: DATE,
        id: ID
        title: "第二级-第二章",
        description: None,
        cover_image: None,
        parent_id: Some(
            1,
        ),
        sort_order: Some(
            2,
        ),
        is_public: None,
        user_id: ID
    },
]
